src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp

```diff
@@ -28,6 +28,8 @@ class Gemm : public ov::op::v0::MatMul {
 
     Gemm(const ov::Output<Node>& A,
          const ov::Output<Node>& B,
+         const std::vector<int64_t>& axes_a,
```

Comment:
I still don't understand why this fusion should be so complicated..
The initial optimization was supposed to cover grouped query attention pattern which is something like KVCache -> Unsqueeze/Reshape -> Broadcast -> Reshape -> Gemm

And as I mentioned in the previous PR, this pattern doesn't require all these axes/target_shape/pattern parameters, only update of Gemm shape inference to allow broadcasts like 2 -> 32. 

Do you have any other patterns besides GQA for this optimization? If no, please simplify that.

Reply:
Simplified UnsqueezeBroadcastReshapeMatmul fusion pass and logic to take only necessary parameters as your suggestion.

Reply:
Is reshape pattern really needed?

Reply:
This is because at least two things are required: the dimension axis to be reshaped and the pattern value of the corresponding index.

Reply:
@andrew-k-park  
Seems that Vladimir means that we can have more strong assumption for the target pattern as only for GQA and MQA. 
So 
1. Fusion pattern & shape inference : We can simplify with additional limitation such as "previous node is KV cache"
2.Remove input pattern and just calculate broadcasted batch dimension (via indirect gemm, no reshape is needed) 
3. However for 2, according to Andrew, indirect gemm is currently only applicable to the beam idx dimension of [0]. Need to fix it to apply for other dim too.

Reply:
Why? I suppose for shape inference it's not needed if you just take out_dim = max(in1_dim, in2_dim) for batch dimensions

Reply:
> @andrew-k-park Seems that Vladimir means that we can have more strong assumption for the target pattern as only for GQA and MQA. So
> 
> 1. Fusion pattern & shape inference : We can simplify with additional limitation such as "previous node is KV cache"
> 2. Remove input pattern and just calculate broadcasted batch dimension (via indirect gemm, no reshape is needed)
> 3. However for 2, according to Andrew, indirect gemm is currently only applicable to the beam idx dimension of [0]. Need to fix it to apply for other dim too.

From my POV we just need to relax broadcasting rules in gemm for batch dimensions. It won't be tightly coupled with KVCache or GQA pattern specifically and may work in other cases. But as long as we don't have a real need to handle more complex broadcast cases, I think there's no need to do this extra processing for reshape/broadcast parameters, as it may impact performance

Reply:
And I thought more about this, we can still handle the broadcasted idx by just dividng the size of corresponding dimension in gemm kernel too, so can apply right now

Reply:
I understand more what you mean. So is it okay to add just boolean type attribute of gemm internal op to distinguish whether or not there is fusion?

Reply:
I think the logic should be generic for all cases (w/ and w/o broadcast fusion). We anyway support broadcasting for batch dimensions, so it's just extension of allowed cases. So it's not clear to me how this boolean flag will be used

Reply:
generalized the logic for all cases (w/ and w/o fusion). @vladimir-paramuzov could you review this requested change again?